In [167]:
import pandas as pd

In [168]:
know_2019 = pd.read_csv('../know_Data/train/KNOW_2019(2) - 복원.csv', encoding='cp949')

In [169]:
know_2019.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8555 entries, 0 to 8554
Data columns (total 153 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int64 
 1    sq1       int64 
 2    sq2       int64 
 3    sq3       int64 
 4    sq4       int64 
 5    sq5       int64 
 6    sq6       int64 
 7    sq7       int64 
 8    sq8       int64 
 9    sq9       int64 
 10   sq10      int64 
 11   sq11      int64 
 12   sq12      int64 
 13   sq13      int64 
 14   sq14      int64 
 15   sq15      int64 
 16   sq16      int64 
 17   kq1_1     int64 
 18   kq1_2     int64 
 19   kq2_1     int64 
 20   kq2_2     int64 
 21   kq3_1     int64 
 22   kq3_2     int64 
 23   kq4_1     int64 
 24   kq4_2     int64 
 25   kq5_1     int64 
 26   kq5_2     int64 
 27   kq6_1     int64 
 28   kq6_2     int64 
 29   kq7_1     int64 
 30   kq7_2     int64 
 31   kq8_1     int64 
 32   kq8_2     int64 
 33   kq9_1     int64 
 34   kq9_2     int64 
 35   kq10_1    int64 
 36   kq10_2    in

In [170]:
know_2019.isnull().sum()[:50]

idx       0
sq1       0
sq2       0
sq3       0
sq4       0
sq5       0
sq6       0
sq7       0
sq8       0
sq9       0
sq10      0
sq11      0
sq12      0
sq13      0
sq14      0
sq15      0
sq16      0
kq1_1     0
kq1_2     0
kq2_1     0
kq2_2     0
kq3_1     0
kq3_2     0
kq4_1     0
kq4_2     0
kq5_1     0
kq5_2     0
kq6_1     0
kq6_2     0
kq7_1     0
kq7_2     0
kq8_1     0
kq8_2     0
kq9_1     0
kq9_2     0
kq10_1    0
kq10_2    0
kq11_1    0
kq11_2    0
kq12_1    0
kq12_2    0
kq13_1    0
kq13_2    0
kq14_1    0
kq14_2    0
kq15_1    0
kq15_2    0
kq16_1    0
kq16_2    0
kq17_1    0
dtype: int64

In [171]:
know_2019.isnull().sum()[50:100]

kq17_2    0
kq18_1    0
kq18_2    0
kq19_1    0
kq19_2    0
kq20_1    0
kq20_2    0
kq21_1    0
kq21_2    0
kq22_1    0
kq22_2    0
kq23_1    0
kq23_2    0
kq24_1    0
kq24_2    0
kq25_1    0
kq25_2    0
kq26_1    0
kq26_2    0
kq27_1    0
kq27_2    0
kq28_1    0
kq28_2    0
kq29_1    0
kq29_2    0
kq30_1    0
kq30_2    0
kq31_1    0
kq31_2    0
kq32_1    0
kq32_2    0
kq33_1    0
kq33_2    0
bq1       0
bq2       0
bq3       0
bq4       0
bq4_1a    0
bq4_1b    0
bq4_1c    0
bq5       0
bq5_1     0
bq5_2     0
bq6       0
bq7       0
bq8_1     0
bq8_2     0
bq8_3     0
bq9       0
bq10      0
dtype: int64

In [172]:
know_2019.isnull().sum()[100:]

bq11         0
bq12_1       0
bq12_2       0
bq12_3       0
bq12_4       0
bq12_5       0
bq13_1       0
bq13_2       0
bq13_3       0
bq14_1       0
bq14_2       0
bq14_3       0
bq14_4       0
bq14_5       0
bq15         0
bq16_1       0
bq16_2       0
bq16_3       0
bq16_4       0
bq16_5       0
bq17         0
bq18_1       0
bq18_2       0
bq18_3       0
bq18_4       0
bq18_5       0
bq18_6       0
bq18_7       0
bq18_8       0
bq18_9       0
bq18_10      3
bq19         0
bq20         0
bq20_1       0
bq21_1       0
bq21_2       0
bq21_3       0
bq22         1
bq23         1
bq24         0
bq25         0
bq26         0
bq27         0
bq27_1       0
bq28         0
bq28_1       0
bq28_2       1
bq29         0
bq30         0
bq31_1       0
bq31_2       0
bq31_3      10
knowcode     0
dtype: int64

In [173]:
for i in know_2019.columns : 
    print(i, len(know_2019[i].value_counts()))

idx 8555
sq1 5
sq2 5
sq3 5
sq4 5
sq5 5
sq6 5
sq7 5
sq8 5
sq9 5
sq10 5
sq11 5
sq12 5
sq13 5
sq14 5
sq15 5
sq16 5
kq1_1 5
kq1_2 8
kq2_1 5
kq2_2 8
kq3_1 5
kq3_2 8
kq4_1 5
kq4_2 8
kq5_1 5
kq5_2 8
kq6_1 5
kq6_2 8
kq7_1 5
kq7_2 8
kq8_1 5
kq8_2 8
kq9_1 5
kq9_2 8
kq10_1 5
kq10_2 8
kq11_1 5
kq11_2 8
kq12_1 5
kq12_2 8
kq13_1 5
kq13_2 8
kq14_1 5
kq14_2 8
kq15_1 5
kq15_2 8
kq16_1 5
kq16_2 8
kq17_1 5
kq17_2 8
kq18_1 5
kq18_2 8
kq19_1 5
kq19_2 8
kq20_1 5
kq20_2 8
kq21_1 5
kq21_2 8
kq22_1 5
kq22_2 8
kq23_1 5
kq23_2 8
kq24_1 5
kq24_2 8
kq25_1 5
kq25_2 8
kq26_1 5
kq26_2 8
kq27_1 5
kq27_2 8
kq28_1 5
kq28_2 8
kq29_1 5
kq29_2 8
kq30_1 5
kq30_2 8
kq31_1 5
kq31_2 8
kq32_1 5
kq32_2 8
kq33_1 5
kq33_2 8
bq1 20
bq2 6
bq3 6
bq4 2
bq4_1a 1559
bq4_1b 718
bq4_1c 287
bq5 2
bq5_1 7
bq5_2 3276
bq6 7
bq7 7
bq8_1 5
bq8_2 5
bq8_3 5
bq9 5
bq10 5
bq11 5
bq12_1 5
bq12_2 6
bq12_3 6
bq12_4 6
bq12_5 5
bq13_1 5
bq13_2 5
bq13_3 5
bq14_1 6
bq14_2 6
bq14_3 6
bq14_4 6
bq14_5 6
bq15 5
bq16_1 5
bq16_2 5
bq16_3 5
bq16_4 5
bq16_5 5
bq1

1. 결측치

- 주관식이라 버릴 항목
bq18_10      3
bq22         1
bq23         1


- 처리해야하는 항목
bq28_2  근로유형      1
bq31_3  근로소득     10


2. 이상치

수작업으로 밀린 것 처리

3. 주관식

- 자격증
bq4_1a 1559
bq4_1b 718
bq4_1c 287

- 버림
bq5_2 3276
bq18_10 2533
bq20_1 6326
bq22 1875
bq23 1716
bq24 603

-학과
bq27_1 1831

4. 숫자여야하는 col
bq26 65
bq30 71
bq31_1 238
bq31_2 161
bq31_3 110


5. 체크해봐야할 col 

bq28 2 & bq29 3

In [174]:
for col in know_2019:
        know_2019[col].replace(' ', 0, inplace = True)

In [175]:
# 주관식문항 

str_li =['bq4_1a','bq4_1b','bq4_1c','bq5_2', 'bq18_10','bq20_1','bq22','bq23','bq24', 'bq27_1']

In [176]:
know_2019 = know_2019.fillna(0)

In [177]:
know_2019.isnull().sum()[100:]

bq11        0
bq12_1      0
bq12_2      0
bq12_3      0
bq12_4      0
bq12_5      0
bq13_1      0
bq13_2      0
bq13_3      0
bq14_1      0
bq14_2      0
bq14_3      0
bq14_4      0
bq14_5      0
bq15        0
bq16_1      0
bq16_2      0
bq16_3      0
bq16_4      0
bq16_5      0
bq17        0
bq18_1      0
bq18_2      0
bq18_3      0
bq18_4      0
bq18_5      0
bq18_6      0
bq18_7      0
bq18_8      0
bq18_9      0
bq18_10     0
bq19        0
bq20        0
bq20_1      0
bq21_1      0
bq21_2      0
bq21_3      0
bq22        0
bq23        0
bq24        0
bq25        0
bq26        0
bq27        0
bq27_1      0
bq28        0
bq28_1      0
bq28_2      0
bq29        0
bq30        0
bq31_1      0
bq31_2      0
bq31_3      0
knowcode    0
dtype: int64

In [178]:
# int형으로 바꾸기 위해서 1.null 값처리 2. bq30 번에 37.5 float 형 처리 필요함 

for i in know_2019.columns : 
    if i == 'bq30':
        know_2019[i] = know_2019[i].astype(float).astype(int)
    elif i not in str_li : 
        know_2019[i] = know_2019[i].astype(int)
    else :
        know_2019[i]=know_2019[i].astype(str)


In [179]:
know_2019.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8555 entries, 0 to 8554
Data columns (total 153 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int32 
 1    sq1       int32 
 2    sq2       int32 
 3    sq3       int32 
 4    sq4       int32 
 5    sq5       int32 
 6    sq6       int32 
 7    sq7       int32 
 8    sq8       int32 
 9    sq9       int32 
 10   sq10      int32 
 11   sq11      int32 
 12   sq12      int32 
 13   sq13      int32 
 14   sq14      int32 
 15   sq15      int32 
 16   sq16      int32 
 17   kq1_1     int32 
 18   kq1_2     int32 
 19   kq2_1     int32 
 20   kq2_2     int32 
 21   kq3_1     int32 
 22   kq3_2     int32 
 23   kq4_1     int32 
 24   kq4_2     int32 
 25   kq5_1     int32 
 26   kq5_2     int32 
 27   kq6_1     int32 
 28   kq6_2     int32 
 29   kq7_1     int32 
 30   kq7_2     int32 
 31   kq8_1     int32 
 32   kq8_2     int32 
 33   kq9_1     int32 
 34   kq9_2     int32 
 35   kq10_1    int32 
 36   kq10_2    in

In [180]:
# _1 에서 1을 선택한 사람중 _2 답변을 한 사람이 있는지 확인

for i in range(33) :
    str5 = 'kq' + str(i+1) +'_1'
    str6 = 'kq' + str(i+1) +'_2'
    
    print(know_2019[(know_2019[str5]==1) & (know_2019[str6]!=0)][str6])
    

Series([], Name: kq1_2, dtype: int32)
Series([], Name: kq2_2, dtype: int32)
Series([], Name: kq3_2, dtype: int32)
Series([], Name: kq4_2, dtype: int32)
Series([], Name: kq5_2, dtype: int32)
Series([], Name: kq6_2, dtype: int32)
Series([], Name: kq7_2, dtype: int32)
Series([], Name: kq8_2, dtype: int32)
Series([], Name: kq9_2, dtype: int32)
Series([], Name: kq10_2, dtype: int32)
Series([], Name: kq11_2, dtype: int32)
Series([], Name: kq12_2, dtype: int32)
Series([], Name: kq13_2, dtype: int32)
Series([], Name: kq14_2, dtype: int32)
Series([], Name: kq15_2, dtype: int32)
Series([], Name: kq16_2, dtype: int32)
Series([], Name: kq17_2, dtype: int32)
Series([], Name: kq18_2, dtype: int32)
Series([], Name: kq19_2, dtype: int32)
Series([], Name: kq20_2, dtype: int32)
Series([], Name: kq21_2, dtype: int32)
Series([], Name: kq22_2, dtype: int32)
Series([], Name: kq23_2, dtype: int32)
Series([], Name: kq24_2, dtype: int32)
Series([], Name: kq25_2, dtype: int32)
Series([], Name: kq26_2, dtype: in

In [181]:
# 답변이 있어야하는 셀 중 빈셀 있는지 확인 

for i in range(33) :
    str5 = 'kq' + str(i+1) +'_1'
    str6 = 'kq' + str(i+1) +'_2'
    
    print(know_2019[(know_2019[str5].isin([2,3,4,5])) & (know_2019[str6]==0)][str6])
    

Series([], Name: kq1_2, dtype: int32)
Series([], Name: kq2_2, dtype: int32)
Series([], Name: kq3_2, dtype: int32)
Series([], Name: kq4_2, dtype: int32)
Series([], Name: kq5_2, dtype: int32)
Series([], Name: kq6_2, dtype: int32)
Series([], Name: kq7_2, dtype: int32)
Series([], Name: kq8_2, dtype: int32)
Series([], Name: kq9_2, dtype: int32)
Series([], Name: kq10_2, dtype: int32)
Series([], Name: kq11_2, dtype: int32)
Series([], Name: kq12_2, dtype: int32)
Series([], Name: kq13_2, dtype: int32)
Series([], Name: kq14_2, dtype: int32)
Series([], Name: kq15_2, dtype: int32)
Series([], Name: kq16_2, dtype: int32)
Series([], Name: kq17_2, dtype: int32)
Series([], Name: kq18_2, dtype: int32)
Series([], Name: kq19_2, dtype: int32)
Series([], Name: kq20_2, dtype: int32)
Series([], Name: kq21_2, dtype: int32)
Series([], Name: kq22_2, dtype: int32)
Series([], Name: kq23_2, dtype: int32)
Series([], Name: kq24_2, dtype: int32)
Series([], Name: kq25_2, dtype: int32)
Series([], Name: kq26_2, dtype: in

In [182]:
# bq27=1(중졸)인 사람중 학과 입력한 사람
know_2019[(know_2019.bq27==1) & (know_2019.bq27_1!='0')]['bq27_1'].value_counts()

없음    12
Name: bq27_1, dtype: int64

In [183]:
null_li5 = know_2019[(know_2019.bq27==1) & (know_2019.bq27_1!='0')].index
null_li5

Int64Index([1058, 1272, 1954, 2320, 2488, 4483, 5108, 5628, 5698, 6199, 7154,
            7583],
           dtype='int64')

In [184]:
for i in null_li5 :
    know_2019.iloc[i,143] = '0'

In [185]:
null_li5 = know_2019[(know_2019.bq27==1) & (know_2019.bq27_1!='0')].index
null_li5

Int64Index([], dtype='int64')

In [186]:
# 고졸 이상중 학과 입력안한 사람 
# 앞에서 '' 을 0 으로 replace했기 때문에 0으로 찾아야함 
know_2019[(know_2019.bq27.isin([2,3,4,5,6])) & (know_2019.bq27_1=='0')]['bq27'].value_counts()

Series([], Name: bq27, dtype: int64)

In [187]:
# 근로시간

null_li = know_2019[know_2019.bq30==0].index
null_li

Int64Index([ 174,  957, 1316, 1451, 1548, 2203, 2543, 4069, 4086, 5144, 5735,
            5766, 6396, 6590, 6694, 6721, 6731, 6850, 7220, 8381],
           dtype='int64')

In [188]:
for i in null_li:
      know_2019.iloc[i,148]=round(know_2019[(know_2019.bq1 ==know_2019.iloc[i,83])&(know_2019.bq30!=0)].bq30.median())

In [189]:
null_li = know_2019[know_2019.bq30==0].index
null_li

Int64Index([], dtype='int64')

In [190]:
null_li2 = know_2019[(know_2019.bq28==1) & (know_2019.bq31_1==0)].index
null_li2

Int64Index([  12,   25,   43,   73,  171,  200,  201,  246,  251,  313,
            ...
            8192, 8233, 8234, 8276, 8286, 8381, 8431, 8467, 8485, 8529],
           dtype='int64', length=148)

In [191]:
for i in null_li2:
      know_2019.iloc[i,149] = round(know_2019[(know_2019.bq1 ==know_2019.iloc[i,83])&(know_2019.bq31_1!=0)].bq31_1.median())

In [192]:
null_li2 = know_2019[(know_2019.bq28==1) & (know_2019.bq31_1==0)].index
null_li2

Int64Index([], dtype='int64')

In [193]:
null_li3 = know_2019[(know_2019.bq28==1) & (know_2019.bq31_2==0)].index
null_li3

Int64Index([  25,   43,   87,  171,  200,  290,  313,  479,  519,  528,
            ...
            8233, 8234, 8276, 8371, 8373, 8374, 8431, 8467, 8485, 8529],
           dtype='int64', length=229)

In [194]:
for i in null_li3:
      know_2019.iloc[i,150] = round(know_2019[(know_2019.bq1 ==know_2019.iloc[i,83])&(know_2019.bq31_2!=0)].bq31_2.median())

In [195]:
null_li3 = know_2019[(know_2019.bq28==1) & (know_2019.bq31_2==0)].index
null_li3

Int64Index([], dtype='int64')

In [196]:
null_li4 = know_2019[(know_2019.bq28==2) & (know_2019.bq31_3==0)].index
null_li4

Int64Index([  52,  174,  367,  965, 1060, 1220, 1244, 1256, 1814, 1885, 1975,
            1997, 2034, 2095, 2203, 2350, 2586, 2646, 2895, 3111, 3165, 3312,
            3558, 3650, 3768, 4208, 4687, 4710, 4974, 5103, 5296, 5321, 5359,
            5449, 5486, 5800, 5889, 6150, 6590, 6669, 6721, 6965, 7097, 7146,
            7220, 7364, 7661, 7748, 7772, 7849, 7886, 8156, 8246],
           dtype='int64')

In [197]:
for i in null_li4:
      know_2019.iloc[i,151] = round(know_2019[(know_2019.bq1 ==know_2019.iloc[i,83])&(know_2019.bq31_3!=0)].bq31_3.median())

In [198]:
null_li4 = know_2019[(know_2019.bq28==2) & (know_2019.bq31_3==0)].index
null_li4

Int64Index([], dtype='int64')

답변이 있어야하는 셀 중 빈셀 있는지 확인 

1. 주관식으로 버릴 항목 
bq18_10 
bq20_1

2. 학과 
bq27_1 bq27 23456인데 학과 공백 많음 
       bq27 1 인 사람 중 없음으로 기재한 사람이 있음 
       
3. 근로시간 
bq30 윗 셀 참조 

4. 근로소득 
종사하고 있는 산업의 평균값으로 처리할 예정 

bq31_1 
bq31_2 
윗셀 참조

bq31_3

In [199]:
know_2019.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8555 entries, 0 to 8554
Data columns (total 153 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int32 
 1    sq1       int32 
 2    sq2       int32 
 3    sq3       int32 
 4    sq4       int32 
 5    sq5       int32 
 6    sq6       int32 
 7    sq7       int32 
 8    sq8       int32 
 9    sq9       int32 
 10   sq10      int32 
 11   sq11      int32 
 12   sq12      int32 
 13   sq13      int32 
 14   sq14      int32 
 15   sq15      int32 
 16   sq16      int32 
 17   kq1_1     int32 
 18   kq1_2     int32 
 19   kq2_1     int32 
 20   kq2_2     int32 
 21   kq3_1     int32 
 22   kq3_2     int32 
 23   kq4_1     int32 
 24   kq4_2     int32 
 25   kq5_1     int32 
 26   kq5_2     int32 
 27   kq6_1     int32 
 28   kq6_2     int32 
 29   kq7_1     int32 
 30   kq7_2     int32 
 31   kq8_1     int32 
 32   kq8_2     int32 
 33   kq9_1     int32 
 34   kq9_2     int32 
 35   kq10_1    int32 
 36   kq10_2    in

In [200]:
know_2019.to_csv("know_2019_enutf.csv", encoding='utf8', index=False)
know_2019.to_csv("know_2019_encp.csv", encoding='cp949', index=False)

In [201]:
df_cp = pd.read_csv("know_2019_encp.csv", encoding='cp949')
df_cp.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8555 entries, 0 to 8554
Data columns (total 153 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int64 
 1    sq1       int64 
 2    sq2       int64 
 3    sq3       int64 
 4    sq4       int64 
 5    sq5       int64 
 6    sq6       int64 
 7    sq7       int64 
 8    sq8       int64 
 9    sq9       int64 
 10   sq10      int64 
 11   sq11      int64 
 12   sq12      int64 
 13   sq13      int64 
 14   sq14      int64 
 15   sq15      int64 
 16   sq16      int64 
 17   kq1_1     int64 
 18   kq1_2     int64 
 19   kq2_1     int64 
 20   kq2_2     int64 
 21   kq3_1     int64 
 22   kq3_2     int64 
 23   kq4_1     int64 
 24   kq4_2     int64 
 25   kq5_1     int64 
 26   kq5_2     int64 
 27   kq6_1     int64 
 28   kq6_2     int64 
 29   kq7_1     int64 
 30   kq7_2     int64 
 31   kq8_1     int64 
 32   kq8_2     int64 
 33   kq9_1     int64 
 34   kq9_2     int64 
 35   kq10_1    int64 
 36   kq10_2    in

In [202]:
know_2019.bq27_1.values

array(['기계공학과', '경영학', '0', ..., '상업과', '기계', '항공정비공학'], dtype=object)